module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
+
+	error = dnode_hold(zfsvfs->z_os, zp->z_id, FTAG, &dn);
+	if (error) {
+		ZFS_EXIT(zfsvfs);
+		return (error);
+	}
+
+	spa = dmu_objset_spa(dn->dn_objset);
+
+	if (fm->fm_flags & FIEMAP_FLAG_SYNC)
+		txg_wait_synced(spa_get_dsl(spa), 0);
+
+	/*
+	 * Lock the entire file against changes while assembling the FIEMAP.
+	 * Then hold open the TXG while generating a map of all pending frees
+	 * and dirty blocks.  This isn't strictly necessary but it is a
+	 * convenient way to determine the range of TXGs to check.
+	 */
+	rl_t *rl = zfs_range_lock(&zp->z_range_lock, 0, UINT64_MAX, RL_READER);
+	open_txg = txg_hold_open(spa_get_dsl(spa), &th);
+	syncing_txg = dirty_txg = spa_syncing_txg(spa);
+
+	(void) dbuf_generate_dirty_maps(dn, fm->fm_dirty_tree,
+	    fm->fm_free_tree, &dirty_txg, open_txg);
+
+	/*
+	 * When the currently syncing TXG could not be checked, likely because
+	 * the dnode was already synced, we need to wait for the syncing TXG
+	 * to fully complete in order to avoid using stale block pointers.
+	 */
+	if (dirty_txg > syncing_txg)
+		txg_wait_synced(spa_get_dsl(spa), syncing_txg);
+
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	mutex_enter(&dn->dn_mtx);
+
+	dnode_phys_t *dnp = dn->dn_phys;
+	fm->fm_file_size = i_size_read(ip);
+	fm->fm_block_size = dnp->dn_datablkszsec << SPA_MINBLOCKSHIFT;
+	fm->fm_fill_count = 0;
+
+	/*
+	 * When there are only pending dirty buffers the block size will not
+	 * yet have been determined.  Assume the maximum block size.
+	 */
+	if (fm->fm_block_size == 0)
+		fm->fm_block_size = zfsvfs->z_max_blksz;
+
+	/*
+	 * When FIEMAP_FLAG_NOMERGE is set and the number of extents for the
+	 * entire file are being requested.  Then in this special case walking
+	 * the entire indirect block tree is not required.
+	 *
+	 * The fill count can be used to determine the number of extents, all
+	 * dirty blocks are assumed to fill holes, and free blocks are assumed
+	 * to overlap with existing free blocks.  This is a safe worst case
+	 * estimate which may slightly over report the number of extents for
+	 * a file being actively overwritten.
+	 *
+	 * Otherwise, the entire block tree needs to be walked to determine
+	 * exactly how the block pointer will be merged.
+	 */
+	if (fm->fm_extents_max == 0 && fm->fm_flags & FIEMAP_FLAG_NOMERGE &&
+	    fm->fm_start == 0 && fm->fm_length == FIEMAP_MAX_OFFSET) {
+		for (int i = 0; i < MIN(dnp->dn_nblkptr, fm->fm_copies); i++)
+			fm->fm_fill_count += BP_GET_FILL(&dnp->dn_blkptr[i]);
+
+		fm->fm_fill_count += P2ROUNDUP(range_tree_space(
+		    fm->fm_dirty_tree), fm->fm_block_size) / fm->fm_block_size;
+	} else {
+		SET_BOOKMARK(&czb, dmu_objset_id(dn->dn_objset),
+		    dn->dn_object, dnp->dn_nlevels - 1, 0);
+
+		for (int i = 0; i < MIN(dnp->dn_nblkptr, fm->fm_copies); i++) {
+			blkptr_t *bp = &dnp->dn_blkptr[i];
+
+			if (BP_GET_FILL(bp) > 0) {
+				czb.zb_blkid = i;
+				error = zfs_fiemap_visit_indirect(spa, dnp, bp,
+				    &czb, zfs_fiemap_cb, (void *)fm);
```

Comment:
This could take arbitrarily long (hours+).  No writes will be processed during that time, since we are holding the txg open.  This seems like a huge performance problem (and denial-of-service attack).  I think we need to rearchitect this to get the info in chunks, dropping locks periodically so that this does not pause the entire storage pool while in progress.

Reply:
Agreed.  The FIEMAP ioctl itself is allowed to return sub-regions of the file if needed (e.g. the extent array is not large enough) and is expected to be re-called until `FIEMAP_EXTENT_LAST` is set, similar to `read()` or `readdir()` needing multiple calls.